test/test_ddr5.py

```diff
@@ -0,0 +1,117 @@
+dq_pattern = partial(test.phy_common.dq_pattern, databits=16, nphases=8, burst=16)
+
+
+class DDR5Tests(unittest.TestCase):
+    SYS_CLK_FREQ = 100e6
+    CMD_LATENCY = 2
+
+    def run_test(self, dut, dfi_sequence, pad_checkers: Mapping[str, Mapping[str, str]], pad_generators=None, **kwargs):
+        # pad_checkers: {clock: {sig: values}}
+        dfi = DFISequencer(dfi_sequence)
+        checkers = {clk: PadChecker(dut.pads, pad_signals) for clk, pad_signals in pad_checkers.items()}
+        generators = defaultdict(list)
+        generators["sys"].append(dfi.generator(dut.dfi))
+        generators["sys"].append(dfi.reader(dut.dfi))
+        for clock, checker in checkers.items():
+            generators[clock].append(checker.run())
+        pad_generators = pad_generators or {}
+        for clock, gens in pad_generators.items():
+            gens = gens if isinstance(gens, list) else [gens]
+            for gen in gens:
+                generators[clock].append(gen(dut.pads))
+        run_simulation(dut, generators, **kwargs)
+        PadChecker.assert_ok(self, checkers)
+        dfi.assert_ok(self)
+
+    def test_ddr5_ca_addressing(self):
+        # Test that bank/address for different commands are correctly serialized to CA pads
+        latency = '00000000' * self.CMD_LATENCY
+        read       = dict(cs_n=0, cas_n=0, ras_n=1, we_n=1, bank=0b101, address=0b1100110000)
+        write_ap   = dict(cs_n=0, cas_n=0, ras_n=1, we_n=0, bank=0b111, address=0b10000000000)
+        activate   = dict(cs_n=0, cas_n=1, ras_n=0, we_n=1, bank=0b010, address=0b11110000111100001)
+        refresh_ab = dict(cs_n=0, cas_n=0, ras_n=0, we_n=1, bank=0b100, address=0)
+        precharge_ab  = dict(cs_n=0, cas_n=1, ras_n=0, we_n=0, bank=0b011, address=0)
+        mrw        = dict(cs_n=0, cas_n=0, ras_n=0, we_n=0, bank=0b110011, address=0b10101010)  # bank=6-bit address, address=8-bit op code
+        zqc_start  = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=0,     address=0b0000101)  # MPC with ZQCAL START operand
+        zqc_latch  = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=0,     address=0b0000100)  # MPC with ZQCAL LATCH operand
+        mrr        = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=1,     address=0b101101)  # 6-bit address (bank=1 selects MRR)
+
+        self.run_test(DDR5SimPHY(sys_clk_freq=self.SYS_CLK_FREQ),
+                      dfi_sequence = [
+                          {0: read, 4: write_ap},
+                          {0: activate, 4: refresh_ab},
+                          {0: precharge_ab, 4: mrw},
+                          {0: zqc_start, 4: zqc_latch},
+                          {0: mrr},
+                      ],
+                      pad_checkers = {"sys8x_90": {
+                          #                 rd     wr       act    ref      pre    mrw      zqcs   zqcl     mrr
+                          'cs':   latency + '0100'+'0100' + '0100'+'0100' + '0100'+'0100' + '0100'+'0100' + '0100'+'0000',
```

Comment:
So, one thing I noticed is actually that for LPDDR4/5 the `cs` signal is not negated on the PADS, whereas for DDR5 the signal should be named `cs_n` rather than `cs`. Moreover, as per the specs, I think that the `cs_n` signal should be always `HIGH` if no command is being executed, and just go to low in case we have a valid command, on the first part of the transaction.

In addition, the commands are sent in SDR (single data rate) meaning that each command transfer takes place in a different clock period, and it can be registered only at the rising edge of the DDR clock

Reply:
I changed the name and updated the values. I also added required changes to litedram related to this signal.